packages/flutter/lib/src/widgets/scroll_activity.dart

```diff
@@ -369,11 +410,57 @@ class ScrollDragController implements Drag {
     }
     if (_reversed) // e.g. an AxisDirection.up scrollable
       offset = -offset;
-    delegate.applyUserOffset(offset);
+
+    // If a scroll frame is still in progress, entering this function means that
+    // an input event is delivered to us too fast. That potentially means a
+    // later event will be too late which could cause the missing of a frame
+    // (https://github.com/flutter/flutter/issues/31086#issuecomment-513044440).
+    // Hence we'll cache it in `_pendingDragDelta` for the next frame to smooth
+    // it out.
+    //
+    // If the input event is sent to us regularly at the same rate of VSYNC (say
+    // at 60Hz, once per 16ms), this would be identical to the old behavior
+    // where `delegate.applyUserOffset(offset)` is always called right away.
+    // That's because `_isScrollFrameInProgress` will always be false at this
+    // point since it will be cleared by the end of a frame (VSYNC) through
+    // `_checkPendingDragCallback`. This is the case for all Android/iOS devices
+    // before iPhone X/XS.
```

Comment:
meta note: this is a commit message, not a comment. The comment should talk about what we do now, not what we used to do. Similarly it shouldn't link to a GitHub comment to explain the situation, it should just explain it inline.